### PR TITLE
Lowpop access levels

### PR DESCRIPTION
**Pull request by EgorDinamit**
## About the Pull Request

Adds "lowpop access" list to jobs and configuration options to enable it.

If population is below certain configurable number - jobs will start with more access, maintenance tunnels by default.

## Why It's Good For The Game

Well, obviously we don't have a lot of players to fill in every job slot required for ship to operate properly.
At this moment only command jobs have full lowpop_access lists, so if anyone wishes to play on ~4 pop they'd probably want to play as CMO/HOS/RD and so on.

## Did you test it?

Yes, and it works.

## Changelog

:cl:
rscadd: Players on lowpop will start with more access. Choose command positions for even more access.
admin: Lowpop access is controlled by config values LOWPOP_ACCESS and LOWPOP_ACCESS_AMOUNT.
/:cl:

### CODE REVIEW

**Review by Shadowtail117 — COMMENTED** [review]
How do you anticipate this affecting a round that gains more population? E.g. going from 3 pop to 15 to more over the course of a round.

**Comment by EgorDinamit** [comment]
That's exactly why this feature is configurable.
If we get to the point where rounds with more than 10 players are not a miracle - we can disable it.
As for the fact that people can late-join? There isn't much we can do about it, see above. We're not yet at a point where people late-join frequently to the point of breaking lowpop.

File: code/game/jobs/job/job.dm [hunk 1]
```diff
@@ -4,6 +4,7 @@
 	var/title = "NOPE"
 	//Job access. The use of minimal_access or access is determined by a config setting: config.jobs_have_minimal_access
 	var/list/minimal_access = list()      // Useful for servers which prefer to only have access given to the places a job absolutely needs (Larger server population)
+	var/list/lowpop_access = list(access_maint_tunnels) // The added access(on-top of minimal or normal access) if server is severely low on population(below ~6, depends on config)
```
Comment (Shadowtail117) [inline code/game/jobs/job/job.dm]:
```suggestion
	var/list/lowpop_access = list(access_maint_tunnels) // The added access (on-top of normal access) if the server is severely low on population, decided by LOWPOP_ACCESS_AMOUNT in the config
```

File: config/example/config.txt [hunk 2]
```diff
@@ -22,6 +22,12 @@ BAN_LEGACY_SYSTEM
 ## Add a # here if you wish to use the setup where jobs have more access. This is intended for servers with low populations - where there are not enough players to fill all roles, so players need to do more than just one job. Also for servers where they don't want people to hide in their own departments.
 JOBS_HAVE_MINIMAL_ACCESS
 
+## Uncomment if you want jobs spawn with more access on lowpop rounds
```
Comment (Shadowtail117) [inline config/example/config.txt]:
```suggestion
## Uncomment if you want jobs to spawn with more access on lowpop rounds

```

File: code/game/jobs/job/job.dm [hunk 3]
```diff
@@ -163,10 +164,14 @@
...
 
 /datum/job/proc/get_access()
+	var/list/returned_access = list()
 	if(minimal_access.len && (!config || config.jobs_have_minimal_access))
-		return src.minimal_access.Copy()
+		returned_access = src.minimal_access.Copy()
 	else
-		return src.access.Copy()
+		returned_access = src.access.Copy()
+	if(lowpop_access.len && (GLOB.clients.len < config.lowpop_access_amount) && (!config || config.lowpop_access))
```
Comment (Shadowtail117) [inline code/game/jobs/job/job.dm]: `&& (!config || config.lowpop_access)`

"...and there is either no config or the config has lowpop_access enabled" -- can you elaborate on what this part of the if statement means?

**Review by Shadowtail117 — DISMISSED** [review]
Also bear in mind that `GLOB.clients.len` will dynamically change to be the number of connected people (including people who have not joined yet), meaning that latejoins above the lowpop amount _will not_ have lowpop access. This, in my opinion, is the optimal solution for the problem I posed, and I have no problem with how it's implemented.

File: code/game/jobs/job/job.dm (hunk 3 above)
Comment (EgorDinamit) [reply code/game/jobs/job/job.dm]: Probably copy-paste from config code, they always use this.
tl;dr, if config somehow doesn't exist - it will assume lowpop_access is enabled.

Comment (Shadowtail117) [reply code/game/jobs/job/job.dm]: If that is the intention, then this would throw an error before that anyway, as trying to access `config.lowpop_access_amount` would result in an NRE.

If the intention is to have it default to on if there is no config, you'll need to rewrite it to short circuit it.

**Review by Shadowtail117 — COMMENTED** [review]
I prefer to note whenever the config is changed, something like:
> admin: Added LOWPOP_ACCESS and LOWPOP_ACCESS_AMOUNT to config.txt. Note that if there is no config, lowpop access will always be in effect.

**Review by Shadowtail117 — APPROVED** [review]
LGTM
